compiler/rustc_trait_selection/src/infer.rs

```diff
@@ -210,13 +211,13 @@ impl<'tcx> OutlivesEnvironmentExt<'tcx> for OutlivesEnvironment<'tcx> {
     fn add_implied_bounds(
         &mut self,
         infcx: &InferCtxt<'a, 'tcx>,
-        fn_sig_tys: &[Ty<'tcx>],
+        fn_sig_tys: FxHashSet<Ty<'tcx>>,
         body_id: hir::HirId,
         span: Span,
     ) {
         debug!("add_implied_bounds()");
 
-        for &ty in fn_sig_tys {
+        for ty in fn_sig_tys {
```

Comment:
This is iterating on a `FxHashSet<Ty<'tcx>>`, which will depend on ASLR (random memory allocation order) and the results could be observable (e.g. potentially in diagnostics order).

I wonder if we should ban iterating `FxHash{Set,Map}` and force the use of `FxIndex{Set,Map}` instead (since presumably the *insertion* order is deterministic).

Reply:
See also https://github.com/rust-lang/rust/issues/63713#issuecomment-685109097 - we should probably use internal lints to force this harder.

Reply:
> we should ban iterating `FxHash{Set,Map}`

YES! I think recommending `FxIndex{Set,Map}` is probably the right choice, though instead of banning it we could use an internal lint whenever using a `FxIndexSet` in a potentially nondet way. Now if the order doesn't matter, e.g. checking for existence without any sideeffects, this lint could be explicitly allowed.

Not sure what's the best way to write such a lint though :sweat_smile: